File: samples/python/mpi/sources/svd_fast_distributed_mpi.py
# file: svd_fast_distributed_mpi.py
#===============================================================================
# Copyright 2017-2019 Intel Corporation.
#
# This software and the related documents are Intel copyrighted  materials,  and
# your use of  them is  governed by the  express license  under which  they were
# provided to you (License).  Unless the License provides otherwise, you may not
# use, modify, copy, publish, distribute,  disclose or transmit this software or
# the related documents without Intel's prior written permission.
#
# This software and the related documents  are provided as  is,  with no express
# or implied  warranties,  other  than those  that are  expressly stated  in the
# License.
#
# License:
# http://software.intel.com/en-us/articles/intel-sample-source-code-license-agr
# eement/
#===============================================================================

#
# !  Content:
# !    Python sample of computing singular value decomposition (SVD) in the
# !    distributed processing mode
# !*****************************************************************************

#
## <a name="DAAL-EXAMPLE-PY-SVD_FAST_DISTRIBUTED_MPI"></a>
## \example svd_fast_distributed_mpi.py
#

import os
import sys
from os.path import join as jp

from mpi4py import MPI
import numpy as np

from daal import step1Local, step2Master, step3Local
from daal.algorithms import svd
from daal.data_management import (
    FileDataSource, InputDataArchive, OutputDataArchive, DataSourceIface, DataCollection
)

utils_folder = os.path.realpath(os.path.abspath(jp(os.environ['DAALROOT'], 'examples', 'python', 'source')))
if utils_folder not in sys.path:
    sys.path.insert(0, utils_folder)
from utils import printNumericTable

# Input data set parameters
nBlocks = 4
MPI_ROOT = 0
DAAL_PREFIX = jp('data', 'distributed')

datasetFileNames = [
    jp(DAAL_PREFIX, 'svd_1.csv'),
    jp(DAAL_PREFIX, 'svd_2.csv'),
    jp(DAAL_PREFIX, 'svd_3.csv'),
    jp(DAAL_PREFIX, 'svd_4.csv')
]

dataFromStep1ForStep3 = None
Sigma = None
V = None
Ui = None

serializedData = None


def computestep1Local():
    global serializedData, dataFromStep1ForStep3

    # Initialize FileDataSource to retrieve the input data from a .csv file
    dataSource = FileDataSource(datasetFileNames[rankId],
                                DataSourceIface.doAllocateNumericTable,
                                DataSourceIface.doDictionaryFromContext)

    # Retrieve the input data
    dataSource.loadDataBlock()

    # Create an algorithm to compute SVD on local nodes
    algorithm = svd.Distributed(step1Local)

    algorithm.input.set(svd.data, dataSource.getNumericTable())

    # Compute SVD
    # OnlinePartialResult class from svd
    pres = algorithm.compute()

    dataFromStep1ForStep2 = pres.get(svd.outputOfStep1ForStep2)
    dataFromStep1ForStep3 = pres.get(svd.outputOfStep1ForStep3)

    # Serialize partial results required by step 2
    dataArch = InputDataArchive()
    dataFromStep1ForStep2.serialize(dataArch)

    nodeResults = dataArch.getArchiveAsArray()

    # Transfer partial results to step 2 on the root node
    serializedData = comm.gather(nodeResults)


def computeOnMasterNode():
    global serializedData, Sigma, V

    # Create an algorithm to compute SVD on the master node
    algorithm = svd.Distributed(step2Master)

    for i in range(nBlocks):
        # Deserialize partial results from step 1
        dataArch = OutputDataArchive(serializedData[i])

        dataForStep2FromStep1 = DataCollection()
        dataForStep2FromStep1.deserialize(dataArch)

        algorithm.input.add(svd.inputOfStep2FromStep1, i, dataForStep2FromStep1)

    # Compute SVD
    # DistributedPartialResult class from svd
    pres = algorithm.compute()

    inputForStep3FromStep2 = pres.getCollection(svd.outputOfStep2ForStep3)

    for i in range(nBlocks):
        # Serialize partial results to transfer to local nodes for step 3
        dataArch = InputDataArchive()
        inputForStep3FromStep2[i].serialize(dataArch)
        length = dataArch.getSizeOfArchive()

        serializedData[i] = np.empty(length, dtype=np.uint8)
        dataArch.copyArchiveToArray(serializedData[i])

    # DistributedPartialResult class from svd
    res = algorithm.getResult()

    Sigma = res.get(svd.singularValues)
    V = res.get(svd.rightSingularMatrix)


def finalizeComputestep1Local():
    global Ui, serializedData

    # Transfer partial results from the root node
    nodeResults = comm.scatter(serializedData)

    # Deserialize partial results from step 2
    dataArch = OutputDataArchive(nodeResults)

    dataFromStep2ForStep3 = DataCollection()
    dataFromStep2ForStep3.deserialize(dataArch)

    # Create an algorithm to compute SVD on the master node
    algorithm = svd.Distributed(step3Local)
    algorithm.input.set(svd.inputOfStep3FromStep1, dataFromStep1ForStep3)
    algorithm.input.set(svd.inputOfStep3FromStep2, dataFromStep2ForStep3)

    # Compute SVD
    algorithm.compute()
    res = algorithm.finalizeCompute()

    Ui = res.get(svd.leftSingularMatrix)

if __name__ == "__main__":

    comm = MPI.COMM_WORLD
    comm_size = comm.Get_size()
    rankId = comm.Get_rank()

    if nBlocks != comm_size:
        if rankId == MPI_ROOT:
            frmt = "{} MPI ranks != {} datasets available, so please start exactly {} ranks."
            print(frmt.format(comm_size, nBlocks, nBlocks))
        sys.exit(0)

    computestep1Local()

    if rankId == MPI_ROOT:
        computeOnMasterNode()

    finalizeComputestep1Local()

    # Print the results
    if rankId == MPI_ROOT:
        printNumericTable(Sigma, "Singular values:")
        printNumericTable(V, "Right orthogonal matrix V:")
        printNumericTable(Ui, "Part of left orthogonal matrix U from root node:", 10)

File: samples/cpp/mpi/sources/svd_fast_distributed_mpi.cpp
/* file: svd_fast_distributed_mpi.cpp */
/*******************************************************************************
* Copyright 2017-2020 Intel Corporation
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*******************************************************************************/

/*
!  Content:
!    C++ sample of computing singular value decomposition (SVD) in the
!    distributed processing mode
!******************************************************************************/

/**
 * <a name="DAAL-EXAMPLE-CPP-SVD_FAST_DISTRIBUTED_MPI"></a>
 * \example svd_fast_distributed_mpi.cpp
 */

#include <mpi.h>
#include "daal.h"
#include "service.h"

using namespace std;
using namespace daal;
using namespace daal::algorithms;

/* Input data set parameters */
const size_t nBlocks = 4;

const string datasetFileNames[] = { "./data/distributed/svd_1.csv", "./data/distributed/svd_2.csv", "./data/distributed/svd_3.csv",
                                    "./data/distributed/svd_4.csv" };

void computestep1Local();
void computeOnMasterNode();
void finalizeComputestep1Local();

int rankId;
int commSize;
#define mpiRoot 0

data_management::DataCollectionPtr dataFromStep1ForStep3;
NumericTablePtr Sigma;
NumericTablePtr V;
NumericTablePtr Ui;

services::SharedPtr<byte> serializedData;
size_t perNodeArchLength;

int main(int argc, char * argv[])
{
    checkArguments(argc, argv, 4, &datasetFileNames[0], &datasetFileNames[1], &datasetFileNames[2], &datasetFileNames[3]);

    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &commSize);
    MPI_Comm_rank(MPI_COMM_WORLD, &rankId);

    if (nBlocks != commSize)
    {
        if (rankId == mpiRoot)
        {
            std::cout << commSize << " MPI ranks != " << nBlocks << " datasets available, so please start exactly " << nBlocks << " ranks.\n" << std::endl;
        }
        MPI_Finalize();
        return 0;
    }

    computestep1Local();

    if (rankId == mpiRoot)
    {
        computeOnMasterNode();
    }

    finalizeComputestep1Local();

    /* Print the results */
    if (rankId == mpiRoot)
    {
        printNumericTable(Sigma, "Singular values:");
        printNumericTable(V, "Right orthogonal matrix V:");
        printNumericTable(Ui, "Part of left orthogonal matrix U from root node:", 10);
    }

    MPI_Finalize();

    return 0;
}

void computestep1Local()
{
    /* Initialize FileDataSource<CSVFeatureManager> to retrieve the input data from a .csv file */
    FileDataSource<CSVFeatureManager> dataSource(datasetFileNames[rankId], DataSource::doAllocateNumericTable, DataSource::doDictionaryFromContext);

    /* Retrieve the input data */
    dataSource.loadDataBlock();

    /* Create an algorithm to compute SVD on local nodes */
    svd::Distributed<step1Local> alg;

    alg.input.set(svd::data, dataSource.getNumericTable());

    /* Compute SVD */
    alg.compute();

    data_management::DataCollectionPtr dataFromStep1ForStep2;
    dataFromStep1ForStep2 = alg.getPartialResult()->get(svd::outputOfStep1ForStep2);
    dataFromStep1ForStep3 = alg.getPartialResult()->get(svd::outputOfStep1ForStep3);

    /* Serialize partial results required by step 2 */
    InputDataArchive dataArch;
    dataFromStep1ForStep2->serialize(dataArch);
    perNodeArchLength = dataArch.getSizeOfArchive();

    /* Serialized data is of equal size on each node if each node called compute() equal number of times */
    if (rankId == mpiRoot)
    {
        serializedData = services::SharedPtr<byte>(new byte[perNodeArchLength * nBlocks]);
    }

    byte * nodeResults = new byte[perNodeArchLength];
    dataArch.copyArchiveToArray(nodeResults, perNodeArchLength);

    /* Transfer partial results to step 2 on the root node */
    MPI_Gather(nodeResults, perNodeArchLength, MPI_CHAR, serializedData.get(), perNodeArchLength, MPI_CHAR, mpiRoot, MPI_COMM_WORLD);

    delete[] nodeResults;
}

void computeOnMasterNode()
{
    /* Create an algorithm to compute SVD on the master node */
    svd::Distributed<step2Master> alg;

    for (size_t i = 0; i < nBlocks; i++)
    {
        /* Deserialize partial results from step 1 */
        OutputDataArchive dataArch(serializedData.get() + perNodeArchLength * i, perNodeArchLength);

        data_management::DataCollectionPtr dataForStep2FromStep1 = data_management::DataCollectionPtr(new data_management::DataCollection());
        dataForStep2FromStep1->deserialize(dataArch);

        alg.input.add(svd::inputOfStep2FromStep1, i, dataForStep2FromStep1);
    }

    /* Compute SVD */
    alg.compute();

    svd::DistributedPartialResultPtr pres            = alg.getPartialResult();
    KeyValueDataCollectionPtr inputForStep3FromStep2 = pres->get(svd::outputOfStep2ForStep3);

    for (size_t i = 0; i < nBlocks; i++)
    {
        /* Serialize partial results to transfer to local nodes for step 3 */
        InputDataArchive dataArch;
        (*inputForStep3FromStep2)[i]->serialize(dataArch);

        if (i == 0)
        {
            perNodeArchLength = dataArch.getSizeOfArchive();
            /* Serialized data is of equal size for each node if it was equal in step 1 */
            serializedData = services::SharedPtr<byte>(new byte[perNodeArchLength * nBlocks]);
        }

        dataArch.copyArchiveToArray(serializedData.get() + perNodeArchLength * i, perNodeArchLength);
    }

    svd::ResultPtr res = alg.getResult();

    Sigma = res->get(svd::singularValues);
    V     = res->get(svd::rightSingularMatrix);
}

void finalizeComputestep1Local()
{
    /* Get the size of the serialized input */
    MPI_Bcast(&perNodeArchLength, sizeof(size_t), MPI_CHAR, mpiRoot, MPI_COMM_WORLD);

    byte * nodeResults = new byte[perNodeArchLength];

    /* Transfer partial results from the root node */

    MPI_Scatter(serializedData.get(), perNodeArchLength, MPI_CHAR, nodeResults, perNodeArchLength, MPI_CHAR, mpiRoot, MPI_COMM_WORLD);

    /* Deserialize partial results from step 2 */
    OutputDataArchive dataArch(nodeResults, perNodeArchLength);

    data_management::DataCollectionPtr dataFromStep2ForStep3 = data_management::DataCollectionPtr(new data_management::DataCollection());
    dataFromStep2ForStep3->deserialize(dataArch);

    delete[] nodeResults;

    /* Create an algorithm to compute SVD on the master node */
    svd::Distributed<step3Local> alg;

    alg.input.set(svd::inputOfStep3FromStep1, dataFromStep1ForStep3);
    alg.input.set(svd::inputOfStep3FromStep2, dataFromStep2ForStep3);

    /* Compute SVD */
    alg.compute();
    alg.finalizeCompute();

    svd::ResultPtr res = alg.getResult();

    Ui = res->get(svd::leftSingularMatrix);
}
